docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,304 @@
+controllers and will require more coordination between developers and users. The second method
+requires less direct intervention from infrastructure providers and puts more resposibility on users, for
+this additional responsibility the users gain immediate access to the feature. These methods are
+mutually exclusive, and the annotations will take preference when specified.
+
+#### Infrastructure Machine Template Status Updates
+
+Infrastructure providers should add a field to the `status` of any Infrastructure Machine Template they reconcile.
+This field will contain the CPU, memory, and GPU resources associated with the machine described by
+the template.  Internally, this field will be represented by a Go `map` type utilizing named constants
+for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values (similar to how resource
+limits and requests are handled for pods).
+
+It is worth mentioning that the Infrastructure Machine Templates are not usually reconciled by themselves.
+Each infrastructure provider will be responsible for determining the best implementation for adding the
+status field based on the information available on their platform.
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU corev1.ResourceName = "cpu"
+    AutoscalerResourceMemory corev1.ResourceName = "memory"
+    AutoscalerResourceGPU corev1.ResourceName = "gpu"
+)
+
+// AutoscalerMachineStatus defines information about individual machines used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineStatus struct {
+    Capacity corev1.ResourceList `json:"capacity,omitempty"`
+}
+
+// AutoscalerMachineTemplateStatus defines information that is used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineTemplateStatus struct {
+    Machine AutoscalerMachineStatus `json:"machine,omitempty"`
+}
+
+// DockerMachineTemplateStatus defines the observed state of a DockerMachineTemplate
+type DockerMachineTemplateStatus struct {
+    Autoscaler AutoscalerMachineTemplateStatus `json:"autoscaler,omitempty"`
+}
+
+// DockerMachineTemplate is the Schema for the dockermachinetemplates API.
+type DockerMachineTemplate struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec DockerMachineTemplateSpec     `json:"spec,omitempty"`
+    Status DockerMachineTemplateStatus `json:"status,omitempty"`
+}
+```
+_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/api/core/v1`_
+
+When used as a manifest, it would look like this:
+
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: DockerMachineTemplate
+metadata:
+  name: workload-md-0
+  namespace: default
+spec:
+  template:
+    spec: {}
+status:
+  autoscaler:
```

Comment:
Do we need the autoscaler prefix here?

Reply:
@vincepri Looking at: https://github.com/kubernetes-sigs/cluster-api/pull/4283/files#r716762327

Does it make sense to drop both the autoscaler and capacity prefix/layer?

Reply:
this structure is what we arrived at after some discussion. i had originally had these values with fewer layers wrapping them, but during the review we added the extra scoping to make it clear. if we want to go back on that i'm ok with, but we should get buy-in from @fabriziopandini as well.

Reply:
I don't have strong opinions either way, so just ignore me :)

Reply:
What about having simply status.capacity=

Reply:
i'm fine with that, i think that brings us back around to a simpler/flatter structure for this. if folks are ok with `status.capacity`, i will adjust the enhancement accordingly.

Reply:
+1 to drop the autoscaler prefix, just to add more color about my comment here and in the annotation suggestion is that in the future we might want to be able to reuse this information or bring it elsewhere as well (machines for example, by inspecting nodes, etc)

Reply:
i'll change the references for the status block and annotations, thanks for the continued discussion.